sdk/core/core/src/az_log.c

```diff
@@ -10,62 +10,71 @@
 #include <az_log_internal.h>
 #include <az_span.h>
 
-#include <stdbool.h>
 #include <stddef.h>
 
 #include <_az_cfg.h>
 
 static az_log_classification const* _az_log_classifications = NULL;
-static size_t _az_log_classifications_length = 0;
-static az_log_fn _az_log_listener = NULL;
+static az_log_message_fn _az_log_message_callback = NULL;
 
-void az_log_set_classifications(
-    az_log_classification const* classifications,
-    size_t classifications_length)
+void az_log_set_classifications(az_log_classification const classifications[])
 {
-  // TODO: thread safety
   _az_log_classifications = classifications;
-  _az_log_classifications_length = classifications_length;
 }
 
-void az_log_set_listener(az_log_fn listener)
+void az_log_set_callback(az_log_message_fn az_log_message_callback)
 {
-  // TODO: thread safety
-  _az_log_listener = listener;
+  _az_log_message_callback = az_log_message_callback;
 }
 
 void az_log_write(az_log_classification classification, az_span message)
 {
-  // TODO: thread safety
-  if (_az_log_listener != NULL && az_log_should_write(classification))
+  if (_az_log_message_callback != NULL && az_log_should_write(classification))
   {
-    (*_az_log_listener)(classification, message);
+    int32_t const length = az_span_length(message);
+    int32_t const capacity = az_span_capacity(message);
+    // Do nothing if span is empty or if truncating it with zero would result in empty span.
+    if (length > 0 && capacity > 1)
+    {
+      int32_t const z_pos = length < capacity ? length + 1 : capacity - 1;
```

Comment:
I think you have a bug here. If length < capacity, then adding 1 to length could set it equal to capacity. And then buf[capacity] =0 (line 46) writes outside the buffer.

I would have just done this:
zero_pos = (length < capacity) ? length : capacity -1;
if (buf[zero_pos != 0) { buf[zero_pos] = 0; }

Reply:
This is why I attempted to change `AZ_SPAN_FROM_STR()` to have capacity that is greater than length by 1. Because the compiler does place the 0 at the end of the string literal, it is there.
But it looks like doing that does break `az_span_slice()`, and so far it is hard for me to get it working in all contexts where it is used with `-1`.
I think I should revert the code pretty much to what I've had before. I would extract adding 0 to a buffer logic into a separate inline function, so that when we update the code that forms the message to be printed, there is less chance to miss/misplace the 0.
I do not feel comfortable appending zero to the span too high in the chain of calls. Yes, it requires only updating code in one place, but now `az_log_wrinte(AZ_SPAN_FROM_STR("Hello"))` would be AV on Windows, and is a UB on microcontrollers, where string would sit in a read-only memory.

Reply:
Maybe the best answer is to put it all back the way it was and pass the span to the customer with the not-0-terminated string and the length.

Reply:
I would keep the `const char*` and `length`, but I would put the code that appends 0 in the same scope where message buffer is created. I would extract the code tat appends 0 to a separate inline internal function that all logging code can reuse.
Requiring `const char*` and `length` in `az_log_write()` + having a zero-terminating function in the logging internal header would also signal to the internal code what is the expected format, and it can use string literals if needed, without trying to modify a read-only memory.

Reply:
@JeffreyRichter , how do you like the current version I just pushed? We still keep `conat chars*` but we also don't try to write `0` to just any buffer. `az_log_write()` would work correctly with the read-only buffer, given that the message is 0-terminated.